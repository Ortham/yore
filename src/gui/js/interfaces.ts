export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface Photo {
  path: string;
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
  height?: number;
  width?: number;
  loaded?: boolean;
}

interface Location {
  Existing?: Coordinates;
  Suggested?: [Coordinates, LocationAccuracy];
}

export interface LocationAccuracy {
  meters: number;
  seconds: number;
}

export interface LocationResponse {
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
}
