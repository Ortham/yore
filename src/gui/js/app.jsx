import 'babel-polyfill';
import 'whatwg-fetch';
import React from 'react';
import ReactDOM from 'react-dom';
import Page from './page';
import { getPhotos, getRootPath } from './requests';

function getInitialState() {
  const state = {
    rootPath: undefined,
    filterPhotos: false,
    currentPhoto: undefined,
    photos: []
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath;

      return getPhotos();
    })
    .then(photos => {
      state.photos = photos;

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(
    <Page rootPath={state.rootPath} photos={state.photos} />,
    document.getElementById('root')
  );
});
