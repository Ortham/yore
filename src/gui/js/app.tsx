import '@babel/polyfill';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import 'whatwg-fetch';
import { Page, PageProps } from './page';
import {
  getInterpolate,
  getLocationHistoryPath,
  getPhotos,
  getRootPath,
  getLocations
} from './requests';
import { updatePhotoLocations } from './photo';

function getInitialState() {
  const state: PageProps = {
    interpolate: false,
    locationHistoryPath: undefined,
    photos: [],
    rootPath: undefined
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath || '';

      return getPhotos();
    })
    .then(photos =>
      getLocations(0, photos.length).then(locations =>
        updatePhotoLocations(photos, locations)
      )
    )
    .then(photos => {
      state.photos = photos;

      return getInterpolate();
    })
    .then(responseBody => {
      state.interpolate = responseBody.interpolate;

      return getLocationHistoryPath();
    })
    .then(responseBody => {
      state.locationHistoryPath = responseBody.locationHistoryPath || '';

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(<Page {...state} />, document.getElementById('root'));
});
