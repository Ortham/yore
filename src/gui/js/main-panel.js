import React from 'react';
import MapArea from './map-area';
import {locationDescription, hasSuggestion} from './photo';

export default function MainPanel(props) {
    if (props.photo) {
        return (
            <main>
                <section>
                    <img src={props.photo.src} />
                    <MapArea photo={props.photo} />
                </section>
                <footer>
                    {locationDescription(props.photo)}
                    <br />
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionApply} >
                        Apply
                    </button>
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionDiscard} >
                        Discard
                    </button>
                </footer>
            </main>
        );
    } else {
        return (
            <main>
                <section>No photo selected.</section>
                <footer>
                    <br />
                    <button disabled>Apply</button>
                    <button disabled>Discard</button>
                </footer>
            </main>
        );
    }
}
