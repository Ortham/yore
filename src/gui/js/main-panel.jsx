import React from 'react';
import PropTypes from 'prop-types';
import MapArea from './map-area';
import { locationDescription, hasSuggestion } from './photo';

export default function MainPanel(props) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}

MainPanel.propTypes = {
  photo: PropTypes.shape({
    src: PropTypes.string
  }),
  handleSuggestionApply: PropTypes.func.isRequired,
  handleSuggestionDiscard: PropTypes.func.isRequired
};

MainPanel.defaultProps = {
  photo: undefined
};
