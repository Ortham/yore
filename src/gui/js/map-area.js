import React from 'react';
import GoogleMapReact from 'google-map-react';
import FaMapMarker from 'react-icons/lib/fa/map-marker';
import {googleMapsCoordinates} from './photo';

function MapMarker(props) {
    return <FaMapMarker lat={props.lat} lng={props.lng} style={{height: '24px', width: '24px', color: 'crimson'}} />;
}

export default function MapArea(props) {
    if (props.photo && props.photo.location) {
        const coord = googleMapsCoordinates(props.photo);
        return (
            <GoogleMapReact center={coord} zoom={5}>
                <MapMarker lat={coord.lat} lng={coord.lng} />
            </GoogleMapReact>
        );
    } else {
        return <GoogleMapReact center={[0, 0]} zoom={5} />;
    }
}
