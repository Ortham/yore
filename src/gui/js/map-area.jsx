import React from 'react';
import PropTypes from 'prop-types';
import GoogleMapReact from 'google-map-react';
import FaMapMarker from 'react-icons/lib/fa/map-marker';
import { googleMapsCoordinates } from './photo';

function MapMarker(props) {
  return (
    <FaMapMarker
      lat={props.lat}
      lng={props.lng}
      style={{
        height: '24px',
        width: '24px',
        color: 'crimson',
        position: 'relative',
        top: '-24px',
        left: '-12px'
      }}
    />
  );
}

MapMarker.propTypes = {
  lat: PropTypes.number.isRequired,
  lng: PropTypes.number.isRequired
};

export default function MapArea(props) {
  if (props.photo && props.photo.location) {
    const coord = googleMapsCoordinates(props.photo);
    return (
      <GoogleMapReact center={coord} zoom={5}>
        <MapMarker lat={coord.lat} lng={coord.lng} />
      </GoogleMapReact>
    );
  }
  return <GoogleMapReact center={[0, 0]} zoom={5} />;
}

MapArea.propTypes = {
  photo: PropTypes.shape({
    location: PropTypes.object
  })
};

MapArea.defaultProps = {
  photo: undefined
};
