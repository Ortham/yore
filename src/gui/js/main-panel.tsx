import * as React from 'react';
import { Photo } from './interfaces';
import { MapArea } from './map-area';
import { hasSuggestion, locationDescription } from './photo';

export interface MainPanelProps {
  photo?: Photo;
  handleSuggestionApply: () => Promise<void>;
  handleSuggestionDiscard: () => void;
}

export function MainPanel(props: MainPanelProps) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}
