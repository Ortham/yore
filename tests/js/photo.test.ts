import * as renderer from 'react-test-renderer';
import { Coordinates, LocationAccuracy } from '../../src/gui/js/interfaces';

jest.mock('react-icons/fi', () => ({
  FiAlertCircle: 'FiAlertCircle',
  FiNavigation: 'FiNavigation',
  FiMapPin: 'FiMapPin'
}));

/* eslint-disable import/first */
import {
  locationDescription,
  hasSuggestion,
  googleMapsCoordinates,
  chooseIcon
} from '../../src/gui/js/photo';
/* eslint-enable import/first */

describe('locationDescription()', () => {
  test('returns suggested location if photo has one', () => {
    const photo = {
      path: '',
      src: '',
      location: {
        Suggested: [
          {},
          {
            meters: 5,
            seconds: 20
          }
        ] as [Coordinates, LocationAccuracy]
      }
    };
    expect(locationDescription(photo)).toBe(
      'Suggested location: accuracy is 5 meters and 20 seconds'
    );
  });

  test('returns existing location if photo has a location but no suggestion', () => {
    const photo = {
      path: '',
      src: '',
      location: {}
    };
    expect(locationDescription(photo)).toBe('Existing location');
  });

  test('returns error if photo has one', () => {
    const photo = {
      path: '',
      src: '',
      error: 'Oh no!'
    };
    expect(locationDescription(photo)).toBe(photo.error);
  });

  test('returns no location if photo has no location or error', () => {
    const photo = {
      path: '',
      src: ''
    };
    expect(locationDescription(photo)).toBe('No location');
  });
});

describe('hasSuggestion()', () => {
  test('returns false if photo has no location', () => {
    const photo = {
      path: '',
      src: ''
    };
    expect(hasSuggestion(photo)).toBe(false);
  });

  test('returns false if photo has a location but no Suggested key', () => {
    const photo = {
      path: '',
      src: '',
      location: {}
    };
    expect(hasSuggestion(photo)).toBe(false);
  });

  test('returns true if photo has a suggested location', () => {
    const photo = {
      path: '',
      src: '',
      location: {
        Suggested: [] as [Coordinates, LocationAccuracy]
      }
    };
    expect(hasSuggestion(photo)).toBe(true);
  });
});

describe('googleMapsCoordinates()', () => {
  test("returns the photo's existing coordinates if it has them", () => {
    const photo = {
      path: '',
      src: '',
      location: {
        Existing: {
          latitude: 52.0,
          longitude: 13.2
        }
      }
    };
    const coordinates = googleMapsCoordinates(photo);
    expect(coordinates.lat).toBe(52.0);
    expect(coordinates.lng).toBe(13.2);
  });

  test("returns the photo's suggested coordinates if it has them", () => {
    const photo = {
      path: '',
      src: '',
      location: {
        Suggested: [
          {
            latitude: 52.0,
            longitude: 13.2
          },
          {
            meters: 5,
            seconds: 20
          }
        ] as [Coordinates, LocationAccuracy]
      }
    };
    const coordinates = googleMapsCoordinates(photo);
    expect(coordinates.lat).toBe(52.0);
    expect(coordinates.lng).toBe(13.2);
  });

  test('returns (0,0) if the photo has no location', () => {
    const photo = {
      path: '',
      src: ''
    };
    const coordinates = googleMapsCoordinates(photo);
    expect(coordinates.lat).toBe(0);
    expect(coordinates.lng).toBe(0);
  });
});

describe('chooseIcon()', () => {
  test('returns an exclamation circle element if the photo has an error', () => {
    const photo = {
      path: '',
      src: '',
      error: 'Oh no!'
    };
    const icon = renderer.create(chooseIcon(photo)).toJSON();
    expect(icon).toMatchSnapshot();
  });

  test('returns a map marker element if the photo has an existing location', () => {
    const photo = {
      path: '',
      src: '',
      location: {
        Existing: {} as Coordinates
      }
    };
    const icon = renderer.create(chooseIcon(photo)).toJSON();
    expect(icon).toMatchSnapshot();
  });

  test('returns a location arrow element if the photo has a suggested location', () => {
    const photo = {
      path: '',
      src: '',
      location: {
        Suggested: [] as [Coordinates, LocationAccuracy]
      }
    };
    const icon = renderer.create(chooseIcon(photo)).toJSON();
    expect(icon).toMatchSnapshot();
  });

  test('returns null if the photo has no error or location', () => {
    const photo = {
      path: '',
      src: ''
    };
    expect(chooseIcon(photo)).toBe(null);
  });
});
