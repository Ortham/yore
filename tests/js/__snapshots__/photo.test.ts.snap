// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chooseIcon() returns a location arrow element if the photo has a suggested location 1`] = `
<div
  className="icon"
>
  <FaLocationArrow
    style={
      Object {
        "left": "-1px",
        "position": "relative",
        "top": "-2px",
      }
    }
  />
</div>
`;

exports[`chooseIcon() returns a map marker element if the photo has an existing location 1`] = `
<div
  className="icon"
>
  <FaMapMarker
    style={
      Object {
        "position": "relative",
        "top": "-2px",
      }
    }
  />
</div>
`;

exports[`chooseIcon() returns an exclamation circle element if the photo has an error 1`] = `
<div
  className="icon"
>
  <FaExclamationCircle
    style={
      Object {
        "position": "relative",
        "top": "-2px",
      }
    }
  />
</div>
`;
